### DISCUSSION TITLE
alm stopped unexpectedly

### DISCUSSION THREAD

**Discussion by whzhangg** [ttadano/alamode#16] [Q&A] [answered]
Hi, I'm doing phonon calculation for GaN and using alm to extract third order force constants. However, the program stopped with error message after some normal output

-----------------------------------------

  LMODEL = least-squares

  Training data file (DFSET) : GaN.DFSET

  NSTART = 1; NEND = 480
  480 entries will be used for training.

  Total Number of Parameters : 2992
  Total Number of Free Parameters : 923
  (stopped)

-----------------------------------------

the error message given is:
terminate called after throwing an instance of 'boost::wrapexcept<std::out_of_range>'

Where should I look for problems?

Training with only the harmonic displacement patterns result in correct phonon dispersion. below is my input settings:

----------------------------------------

General:
  PREFIX = GaN
  MODE = optimize
  NAT = 256; NKD = 2
  PRINTSYM = 1; TOLERANCE = 0.01
  KD =   Ga   N
  PERIODIC =   1  1  1
  MAGMOM = 

 Interaction:
  NORDER = 2
  NBODY =   2  3

 Optimize:
  LMODEL = least-squares
  DFSET = GaN.DFSET
  NDATA = 480; NSTART = 1; NEND = 480   SKIP = 

  ICONST = 11
  ROTAXIS = 
  FC2XML = 
  FC3XML = 

  SPARSE = 0
  SPARSESOLVER = SimplicialLDLT
  CONV_TOL = 1e-08
  MAXITER = 10000

----------------------------

Thank you

**Comment by ttadano** [comment, marked as answer]
Hi,
I have not encountered that error so far.
Could you also put the &cell and &position fields of the ALM input here?
That would help identify the reason.

**Reply by whzhangg** [reply]
Thank you for the reply.

I paste the input below, but I also made a onedrive link where I upload the force set and in/out put to reproduce to problem https://1drv.ms/u/s!AuIRAzAWmZn_gfA6hSCWStaj3ZPlzg?e=ImRaGo

---------------------------------
4*4*4 of the primitive cell

&cell
1.88973  #from A to Bohr
  12.64480844   0.00000000   0.00000000
  -6.32240420  10.95072532   0.00000000
   0.00000000   0.00000000  20.61974012
/

(total  256 atoms)
&position
  1      0.08333333      0.16666667      0.12532337
  1      0.16666667      0.08333333      0.00032337
  2      0.08333333      0.16666667      0.21940163
  2      0.16666667      0.08333333      0.09440163
  1      0.08333333      0.16666667      0.37532337
  1      0.16666667      0.08333333      0.25032337
  2      0.08333333      0.16666667      0.46940163
  2      0.16666667      0.08333333      0.34440163
  1      0.08333333      0.16666667      0.62532337
  1      0.16666667      0.08333333      0.50032337
  2      0.08333333      0.16666667      0.71940163
  2      0.16666667      0.08333333      0.59440163
  1      0.08333333      0.16666667      0.87532337
  1      0.16666667      0.08333333      0.75032337
  2      0.08333333      0.16666667      0.96940163
  2      0.16666667      0.08333333      0.84440163
  1      0.08333333      0.41666667      0.12532337
  1      0.16666667      0.33333333      0.00032337
  2      0.08333333      0.41666667      0.21940163
  2      0.16666667      0.33333333      0.09440163
  1      0.08333333      0.41666667      0.37532337
  1      0.16666667      0.33333333      0.25032337
  2      0.08333333      0.41666667      0.46940163
  2      0.16666667      0.33333333      0.34440163
  1      0.08333333      0.41666667      0.62532337
  1      0.16666667      0.33333333      0.50032337
  2      0.08333333      0.41666667      0.71940163
  2      0.16666667      0.33333333      0.59440163
  1      0.08333333      0.41666667      0.87532337
  1      0.16666667      0.33333333      0.75032337
  2      0.08333333      0.41666667      0.96940163
  2      0.16666667      0.33333333      0.84440163
  1      0.08333333      0.66666667      0.12532337
  1      0.16666667      0.58333333      0.00032337
  2      0.08333333      0.66666667      0.21940163
  2      0.16666667      0.58333333      0.09440163
  1      0.08333333      0.66666667      0.37532337
  1      0.16666667      0.58333333      0.25032337
  2      0.08333333      0.66666667      0.46940163
  2      0.16666667      0.58333333      0.34440163
  1      0.08333333      0.66666667      0.62532337
  1      0.16666667      0.58333333      0.50032337
  2      0.08333333      0.66666667      0.71940163
  2      0.16666667      0.58333333      0.59440163
  1      0.08333333      0.66666667      0.87532337
  1      0.16666667      0.58333333      0.75032337
  2      0.08333333      0.66666667      0.96940163
  2      0.16666667      0.58333333      0.84440163
  1      0.08333333      0.91666667      0.12532337
  1      0.16666667      0.83333333      0.00032337
  2      0.08333333      0.91666667      0.21940163
  2      0.16666667      0.83333333      0.09440163
  1      0.08333333      0.91666667      0.37532337
  1      0.16666667      0.83333333      0.25032337
  2      0.08333333      0.91666667      0.46940163
  2      0.16666667      0.83333333      0.34440163
  1      0.08333333      0.91666667      0.62532337
  1      0.16666667      0.83333333      0.50032337
  2      0.08333333      0.91666667      0.71940163
  2      0.16666667      0.83333333      0.59440163
  1      0.08333333      0.91666667      0.87532337
  1      0.16666667      0.83333333      0.75032337
  2      0.08333333      0.91666667      0.96940163
  2      0.16666667      0.83333333      0.84440163
  1      0.33333333      0.16666667      0.12532337
  1      0.41666667      0.08333333      0.00032337
  2      0.33333333      0.16666667      0.21940163
  2      0.41666667      0.08333333      0.09440163
  1      0.33333333      0.16666667      0.37532337
  1      0.41666667      0.08333333      0.25032337
  2      0.33333333      0.16666667      0.46940163
  2      0.41666667      0.08333333      0.34440163
  1      0.33333333      0.16666667      0.62532337
  1      0.41666667      0.08333333      0.50032337
  2      0.33333333      0.16666667      0.71940163
  2      0.41666667      0.08333333      0.59440163
  1      0.33333333      0.16666667      0.87532337
  1      0.41666667      0.08333333      0.75032337
  2      0.33333333      0.16666667      0.96940163
  2      0.41666667      0.08333333      0.84440163
  1      0.33333333      0.41666667      0.12532337
  1      0.41666667      0.33333333      0.00032337
  2      0.33333333      0.41666667      0.21940163
  2      0.41666667      0.33333333      0.09440163
  1      0.33333333      0.41666667      0.37532337
  1      0.41666667      0.33333333      0.25032337
  2      0.33333333      0.41666667      0.46940163
  2      0.41666667      0.33333333      0.34440163
  1      0.33333333      0.41666667      0.62532337
  1      0.41666667      0.33333333      0.50032337
  2      0.33333333      0.41666667      0.71940163
  2      0.41666667      0.33333333      0.59440163
  1      0.33333333      0.41666667      0.87532337
  1      0.41666667      0.33333333      0.75032337
  2      0.33333333      0.41666667      0.96940163
  2      0.41666667      0.33333333      0.84440163
  1      0.33333333      0.66666667      0.12532337
  1      0.41666667      0.58333333      0.00032337
  2      0.33333333      0.66666667      0.21940163
  2      0.41666667      0.58333333      0.09440163
  1      0.33333333      0.66666667      0.37532337
  1      0.41666667      0.58333333      0.25032337
  2      0.33333333      0.66666667      0.46940163
  2      0.41666667      0.58333333      0.34440163
  1      0.33333333      0.66666667      0.62532337
  1      0.41666667      0.58333333      0.50032337
  2      0.33333333      0.66666667      0.71940163
  2      0.41666667      0.58333333      0.59440163
  1      0.33333333      0.66666667      0.87532337
  1      0.41666667      0.58333333      0.75032337
  2      0.33333333      0.66666667      0.96940163
  2      0.41666667      0.58333333      0.84440163
  1      0.33333333      0.91666667      0.12532337
  1      0.41666667      0.83333333      0.00032337
  2      0.33333333      0.91666667      0.21940163
  2      0.41666667      0.83333333      0.09440163
  1      0.33333333      0.91666667      0.37532337
  1      0.41666667      0.83333333      0.25032337
  2      0.33333333      0.91666667      0.46940163
  2      0.41666667      0.83333333      0.34440163
  1      0.33333333      0.91666667      0.62532337
  1      0.41666667      0.83333333      0.50032337
  2      0.33333333      0.91666667      0.71940163
  2      0.41666667      0.83333333      0.59440163
  1      0.33333333      0.91666667      0.87532337
  1      0.41666667      0.83333333      0.75032337
  2      0.33333333      0.91666667      0.96940163
  2      0.41666667      0.83333333      0.84440163
  1      0.58333333      0.16666667      0.12532337
  1      0.66666667      0.08333333      0.00032337
  2      0.58333333      0.16666667      0.21940163
  2      0.66666667      0.08333333      0.09440163
  1      0.58333333      0.16666667      0.37532337
  1      0.66666667      0.08333333      0.25032337
  2      0.58333333      0.16666667      0.46940163
  2      0.66666667      0.08333333      0.34440163
  1      0.58333333      0.16666667      0.62532337
  1      0.66666667      0.08333333      0.50032337
  2      0.58333333      0.16666667      0.71940163
  2      0.66666667      0.08333333      0.59440163
  1      0.58333333      0.16666667      0.87532337
  1      0.66666667      0.08333333      0.75032337
  2      0.58333333      0.16666667      0.96940163
  2      0.66666667      0.08333333      0.84440163
  1      0.58333333      0.41666667      0.12532337
  1      0.66666667      0.33333333      0.00032337
  2      0.58333333      0.41666667      0.21940163
  2      0.66666667      0.33333333      0.09440163
  1      0.58333333      0.41666667      0.37532337
  1      0.66666667      0.33333333      0.25032337
  2      0.58333333      0.41666667      0.46940163
  2      0.66666667      0.33333333      0.34440163
  1      0.58333333      0.41666667      0.62532337
  1      0.66666667      0.33333333      0.50032337
  2      0.58333333      0.41666667      0.71940163
  2      0.66666667      0.33333333      0.59440163
  1      0.58333333      0.41666667      0.87532337
  1      0.66666667      0.33333333      0.75032337
  2      0.58333333      0.41666667      0.96940163
  2      0.66666667      0.33333333      0.84440163
  1      0.58333333      0.66666667      0.12532337
  1      0.66666667      0.58333333      0.00032337
  2      0.58333333      0.66666667      0.21940163
  2      0.66666667      0.58333333      0.09440163
  1      0.58333333      0.66666667      0.37532337
  1      0.66666667      0.58333333      0.25032337
  2      0.58333333      0.66666667      0.46940163
  2      0.66666667      0.58333333      0.34440163
  1      0.58333333      0.66666667      0.62532337
  1      0.66666667      0.58333333      0.50032337
  2      0.58333333      0.66666667      0.71940163
  2      0.66666667      0.58333333      0.59440163
  1      0.58333333      0.66666667      0.87532337
  1      0.66666667      0.58333333      0.75032337
  2      0.58333333      0.66666667      0.96940163
  2      0.66666667      0.58333333      0.84440163
  1      0.58333333      0.91666667      0.12532337
  1      0.66666667      0.83333333      0.00032337
  2      0.58333333      0.91666667      0.21940163
  2      0.66666667      0.83333333      0.09440163
  1      0.58333333      0.91666667      0.37532337
  1      0.66666667      0.83333333      0.25032337
  2      0.58333333      0.91666667      0.46940163
  2      0.66666667      0.83333333      0.34440163
  1      0.58333333      0.91666667      0.62532337
  1      0.66666667      0.83333333      0.50032337
  2      0.58333333      0.91666667      0.71940163
  2      0.66666667      0.83333333      0.59440163
  1      0.58333333      0.91666667      0.87532337
  1      0.66666667      0.83333333      0.75032337
  2      0.58333333      0.91666667      0.96940163
  2      0.66666667      0.83333333      0.84440163
  1      0.83333333      0.16666667      0.12532337
  1      0.91666667      0.08333333      0.00032337
  2      0.83333333      0.16666667      0.21940163
  2      0.91666667      0.08333333      0.09440163
  1      0.83333333      0.16666667      0.37532337
  1      0.91666667      0.08333333      0.25032337
  2      0.83333333      0.16666667      0.46940163
  2      0.91666667      0.08333333      0.34440163
  1      0.83333333      0.16666667      0.62532337
  1      0.91666667      0.08333333      0.50032337
  2      0.83333333      0.16666667      0.71940163
  2      0.91666667      0.08333333      0.59440163
  1      0.83333333      0.16666667      0.87532337
  1      0.91666667      0.08333333      0.75032337
  2      0.83333333      0.16666667      0.96940163
  2      0.91666667      0.08333333      0.84440163
  1      0.83333333      0.41666667      0.12532337
  1      0.91666667      0.33333333      0.00032337
  2      0.83333333      0.41666667      0.21940163
  2      0.91666667      0.33333333      0.09440163
  1      0.83333333      0.41666667      0.37532337
  1      0.91666667      0.33333333      0.25032337
  2      0.83333333      0.41666667      0.46940163
  2      0.91666667      0.33333333      0.34440163
  1      0.83333333      0.41666667      0.62532337
  1      0.91666667      0.33333333      0.50032337
  2      0.83333333      0.41666667      0.71940163
  2      0.91666667      0.33333333      0.59440163
  1      0.83333333      0.41666667      0.87532337
  1      0.91666667      0.33333333      0.75032337
  2      0.83333333      0.41666667      0.96940163
  2      0.91666667      0.33333333      0.84440163
  1      0.83333333      0.66666667      0.12532337
  1      0.91666667      0.58333333      0.00032337
  2      0.83333333      0.66666667      0.21940163
  2      0.91666667      0.58333333      0.09440163
  1      0.83333333      0.66666667      0.37532337
  1      0.91666667      0.58333333      0.25032337
  2      0.83333333      0.66666667      0.46940163
  2      0.91666667      0.58333333      0.34440163
  1      0.83333333      0.66666667      0.62532337
  1      0.91666667      0.58333333      0.50032337
  2      0.83333333      0.66666667      0.71940163
  2      0.91666667      0.58333333      0.59440163
  1      0.83333333      0.66666667      0.87532337
  1      0.91666667      0.58333333      0.75032337
  2      0.83333333      0.66666667      0.96940163
  2      0.91666667      0.58333333      0.84440163
  1      0.83333333      0.91666667      0.12532337
  1      0.91666667      0.83333333      0.00032337
  2      0.83333333      0.91666667      0.21940163
  2      0.91666667      0.83333333      0.09440163
  1      0.83333333      0.91666667      0.37532337
  1      0.91666667      0.83333333      0.25032337
  2      0.83333333      0.91666667      0.46940163
  2      0.91666667      0.83333333      0.34440163
  1      0.83333333      0.91666667      0.62532337
  1      0.91666667      0.83333333      0.50032337
  2      0.83333333      0.91666667      0.71940163
  2      0.91666667      0.83333333      0.59440163
  1      0.83333333      0.91666667      0.87532337
  1      0.91666667      0.83333333      0.75032337
  2      0.83333333      0.91666667      0.96940163
  2      0.91666667      0.83333333      0.84440163
/

**Reply by ttadano** [reply]
Thank you!

This error occurred probably because the number of numerical digits for the &cell and &potision fields were not enough. Currently, you are using 8 digits, for example, 0.83333333, but I would suggest using 15 digits to avoid this error. 

Here are several ways to solve the issue:

1. Increase the numerical digits for the &cell and &position fields. This will also improve the accuracy of third-order force constants. So, I recommend this option.
2. Install the develop branch of alm and set ``FC_BASIS = Lattice``. This version is more robust against slightly distorted input structures.
3. Set a larger value of ``TOL_CONST``, say ``TOL_CONST = 1.0e-4``, in the &general field.

**Reply by whzhangg** [reply]
Thank you, I found that using the develop branch directly solve the issue and I can optimize 3rd force constant with the original input. But I will also increase the digits and compare the difference.
